### DISCUSSION TITLE
Smart-contract account plugins -- List you ideas!

### DISCUSSION THREAD

**Discussion by imqdee** [LedgerHQ/ledger-fresh-management#60] [Ideas]
In this discussion, we'll list all our plugin ideas.

**Comment by imqdee** [comment]
Reminder here's the list of plugin we plan developing for the initial version of Ledger Fresh:
- Session keys
- WebAuthn verification
- Multi-signer management (Nano / Webauthn...)

Here's a non-exhaustive list of plugins we have in mind for the future: 
- Heir system
- Treshold

**Comment by CremaFR** [comment]
One "easy" day one plugin could be Ledger Protect. I had some notes about it that I will paste here

**Ledger Protect**

This plugin will act as a solution to recover your account in case you lost your nano and your 24 words backup. 
This plugin would be added by an user on its wallet and give access to Ledger to allow the company to change the admin signer of your account. See this as a forgot password option.
Some business rules about this plugin are:
The change of the admin will take a few days if not weeks before being effective to prevent thief
The plugin can be removed at any time, by removing the plugin the admin signer can’t be changed by Ledger anymore
The plugin might not be free, i.e the plugin could be allowed to transfer ERC20 for its payment if done onchain.
Note: The plugin could be use to give this same access rules to anyone in case you don't trust Ledger/want self custody/have a backup nano and act as a "guardian" in Argent's terms.

Technically this would mean changing the public key associated to the admin plugin.

**Comment by CremaFR** [comment]
I would also add to the list 

**Admin Plugin**

Or in a way default plugin? A plugin that **should** use a nano as a public key. Only requires a signature validation and has no restriction to execute functions on the contract (it can add/remove plugins, update w/o timelock parameters from the account)

**Comment by CremaFR** [comment]
**Ledger Chest**

Ledger Chest's Idea is to add a way to protect your account from malicious transactions using a _proof of storage_ of the assets your want to protect. This will make sure at the end of a transaction, the state hasn't changed and your funds are still present.



Challenges and open questions:

- Check gas consumption, as many state reads could be expensive.
- It protects against transactions that transfer directly the asset, what about:
    - `approve` and `approveAll` ? Check the storage value of it too?
    - permit and signature? [EIP-2612](https://eips.ethereum.org/EIPS/eip-2612) and now [permit2](https://docs.uniswap.org/contracts/permit2/reference/signature-transfer) that allows signature transfer for any ERC-20. 

Random notes:

- Signature of proof of storage should be signed with the TxHash or something link to the tx to not be replayable, if passed as a plugin. This requires a change in  `validate` function for **all** plugin, as this change the signature. Using only the pluginId is not a good option here, as the storage value could be swapped

**Comment by rdubois-crypto** [comment]
Ledger Lightning

This pluggins enables the creation of a starknet lightning channel between two participants using starknet tokens/ERC20. It uses a Musig 2 and adaptator over the stark curve and using HashPedersen as hashing function. Note that it only implements a 1O1 channel, or limited sized network (a few friends doing a tricount) not a whole network.

**Comment by rdubois-crypto** [comment]
Ledger PQ

This pluggins enables to secure transaction using a post quantum signature (FALCON512) to secure the transactions, for those fearing the post quantum apocalypse.

**Comment by rdubois-crypto** [comment]
Ledger Tornado

An implementation of a Starknet Mixer using RingCT2/Triptych anonymization algorithm.
